java/src/org/openqa/selenium/remote/BiDiCommandExecutor.java

```diff
@@ -0,0 +1,167 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.remote;
+
+import static org.openqa.selenium.remote.DriverCommand.GET;
+import static org.openqa.selenium.remote.DriverCommand.PRINT_PAGE;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import org.openqa.selenium.UnsupportedCommandException;
+import org.openqa.selenium.bidi.Network;
+import org.openqa.selenium.bidi.Script;
+import org.openqa.selenium.bidi.browsingcontext.BrowsingContext;
+import org.openqa.selenium.bidi.browsingcontext.ReadinessState;
+import org.openqa.selenium.json.Json;
+import org.openqa.selenium.json.JsonInput;
+import org.openqa.selenium.json.TypeToken;
+import org.openqa.selenium.print.PageMargin;
+import org.openqa.selenium.print.PageSize;
+import org.openqa.selenium.print.PrintOptions;
+
+public class BiDiCommandExecutor implements CommandExecutor {
+
+  private static Json JSON = new Json();
+
+  private final RemoteWebDriver driver;
+
+  private final AtomicReference<BrowsingContext> currentContext = new AtomicReference<>();
+  private final AtomicReference<Script> script = new AtomicReference<>();
+  private final AtomicReference<Network> network = new AtomicReference<>();
+
+  private final Map<String, BiFunction<Command, RemoteWebDriver, Response>> commandHandlerMap =
+      new HashMap<>();
+
+  // Each browsing context has an associated id, maintains state
+  // We will need to maintain a map of all the browsingContext to run further commands
+  // Switching between tabs etc might be tricky
+  private final Map<String, BrowsingContext> browsingContextMap = new HashMap<>();
+
+  public BiDiCommandExecutor(RemoteWebDriver driver) {
+    this.driver = driver;
+    init(this.driver);
+  }
+
+  private void init(RemoteWebDriver driver) {
+    script.set(new Script(driver));
+    network.set(new Network(driver));
+
+    BrowsingContext parentContext = new BrowsingContext(driver, driver.getWindowHandle());
+    browsingContextMap.put(parentContext.getId(), parentContext);
+    currentContext.set(parentContext);
+
+    commandHandlerMap.put(
+        GET,
+        (command, webDriver) -> {
+          String pageLoadStrategy =
+              (String) webDriver.getCapabilities().getCapability("pageLoadStrategy");
+          currentContext
+              .get()
+              .navigate(
+                  (String) command.getParameters().get("url"),
+                  ReadinessState.getReadinessState(pageLoadStrategy));
+          return new Response(webDriver.getSessionId());
+        });
```

Comment:
Should we define all these parameters somewhere else? I assume we don't want to declare them all here, as this is a PoC.

Reply:
I did think of using a codec but that did not fit properly here. I do want to populate the command map in some other place. I will look into it. You are right, executor is not the proper place for it.

Reply:
That is fine. As a first step I do not mind having it here, but it is good we want to have it somewhere else as we move forward.

Reply:
+1 I have started working on it too.

Reply:
I added a BiDiCommandMapper. I think that might help.